include/merlin/array_kernels.cuh

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http:///www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cooperative_groups.h>
+#include "cuda_runtime.h"
+#include "thrust/device_vector.h"
+#include "thrust/execution_policy.h"
+#include "thrust/scan.h"
+#include "types.cuh"
+#include "utils.cuh"
+
+namespace nv {
+namespace merlin {
+
+template <typename T>
+__global__ void broadcast(T* data, T val, size_t n) {
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  if (tid < n) {
+    data[tid] = val;
+  }
+}
+
+template <typename K>
+__global__ void keys_not_empty(const K* keys, bool* masks, size_t n) {
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  if (tid < n) {
+    masks[tid] = keys[tid] == EMPTY_KEY ? true : false;
+  }
+}
+
+template <typename Tidx, int TILE_SIZE = 8>
+__global__ void gpu_cell_count(const bool* masks, Tidx* offsets, size_t n,
+                               size_t* n_existed) {
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  auto g = cg::tiled_partition<TILE_SIZE>(cg::this_thread_block());
+  int rank = g.thread_rank();
+  bool is_existed = false;
+  if (tid < n) {
+    if (masks[tid]) {
+      is_existed = true;
+    }
+  }
+  unsigned int vote = g.ballot(is_existed);
+  int g_ones = __popc((int)vote);
+  if (rank == 0 && tid < n) {
+    offsets[tid / TILE_SIZE] = static_cast<Tidx>(g_ones);
+    atomicAdd(static_cast<uint64_t*>(n_existed), static_cast<uint64_t>(g_ones));
```

Comment:
In my experience, the `cuda::atomic<size_t> fetch_add` will has better performance than `atomicAdd`, but it depends on the profiling result.